File: CSharp/Ops/Archiver/OPSArchiverIn.cs
///////////////////////////////////////////////////////////
//  OPSArchiverIn.cs
//  Implementation of the Class OPSArchiverIn
//  Created on:      12-nov-2011 09:25:32
//  Author:
///////////////////////////////////////////////////////////

using System;
using System.Collections;
using System.Collections.Generic;

namespace Ops 
{
	public class OPSArchiverIn : IArchiverInOut 
    {
		private SerializableCompositeFactory compositeFactory;
		private ReadByteBuffer readBuf;

		/// 
		/// <param name="buf"></param>
		public OPSArchiverIn(ReadByteBuffer buf)
        {
            compositeFactory = OPSObjectFactory.GetInstance();
            readBuf = buf;
		}

        /// <summary>
        /// 
        /// </summary>
        /// <param name="buf"></param>
        /// <param name="factory"></param>
        public OPSArchiverIn(ReadByteBuffer buf, OPSObjectFactory factory)
        {
            compositeFactory = factory;
            readBuf = buf;
        }

        public override bool IsOut()
        {
            return false;
        }

		/// 
		/// <param name="e"></param>
		public void Add(ISerializableFactory item)
        {
            compositeFactory.Add(item);
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override int Inout(string name, int v)
        {
            return readBuf.ReadInt();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override long Inout(string name, long v)
        {
            return readBuf.ReadLong();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override byte Inout(string name, byte v)
        {
            return readBuf.ReadByte();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override short Inout(string name, short v)
        {
            return readBuf.ReadShort();
		}

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override float Inout(string name, float v)
        {
            return readBuf.ReadFloat();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override bool Inout(string name, bool v)
        {
            return readBuf.ReadBoolean();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override string Inout(string name, string v)
        {
            return readBuf.ReadString();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override double Inout(string name, double v)
        {
            return readBuf.ReadDouble();
        }

        /// 
        /// <param name="name"></param>
        /// <param name="v"></param>
        public override ISerializable Inout(string name, ISerializable v)
        {
            string types = readBuf.ReadString();
            ISerializable newSer = compositeFactory.Create(types);
            if (newSer != null)
            {
                //Do this to preserve type information even if slicing has occured.
                (newSer as OPSObject).SetTypesString(types);

                newSer.Serialize(this);
            }
            return newSer;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<bool> InoutBooleanList(string name, List<bool> v)
        {
            return readBuf.ReadBooleanArr();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<byte> InoutByteList(string name, List<byte> v)
        {
            return readBuf.ReadByteArr();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<double> InoutDoubleList(string name, List<double> v)
        {
            return readBuf.ReadDoubleArr();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<float> InoutFloatList(string name, List<float> v)
        {
            return readBuf.ReadFloatArr();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<int> InoutIntegerList(string name, List<int> v)
        {
            return readBuf.ReadIntArr();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<long> InoutLongList(string name, List<long> v)
        {
            return readBuf.ReadLongArr();
        }

        // NB! we assume that the object is a List<X> where X implements ISerializable.
        public override IList InoutSerializableList(string name, IList v)
        {
            Type type = v.GetType().GetGenericArguments()[0];
            IList list = (IList)Activator.CreateInstance((typeof(List<>).MakeGenericType(type)));
            int size = readBuf.ReadInt();
            for (int i = 0; i < size; i++)
            {
                list.Add(Inout("", (ISerializable)null));
            }
            return list;
        }

        // NB! we assume that the object is a List<T> where T implements ISerializable.
        // This method skips the factory for creating the elements 
        public override IList InoutSerializableList<T>(string name, IList v)
        {
            Type type = v.GetType().GetGenericArguments()[0];
            IList list = (IList)Activator.CreateInstance((typeof(List<>).MakeGenericType(type)));
            int size = readBuf.ReadInt();
            for (int i = 0; i < size; i++)
            {
                // For each element
                readBuf.ReadString();
                ISerializable obj = (ISerializable)Activator.CreateInstance<T>();
                obj.Serialize(this);
                list.Add(obj);
            }
            return list;
        }

        /// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<short> InoutShortList(string name, List<short> v)
        {
            return readBuf.ReadShortArr();
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public override List<string> InoutStringList(string name, List<string> v)
        {
            return readBuf.ReadStringArr();
        }

		/// 
		/// <param name="o"></param>
        public void remove(ISerializableFactory item)
        {
            compositeFactory.Remove(item);
        }

	}
}

File: Java/OPSJLib/src/ops/archiver/OPSArchiverIn.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
* Copyright (C) 2019 Lennart Andersson.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops.archiver;

import configlib.ArchiverInOut;
import configlib.Serializable;
import configlib.SerializableCompositeFactory;
import configlib.SerializableFactory;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Vector;
import ops.ReadByteBuffer;

/**
 *
 * @author angr
 */
public class OPSArchiverIn extends ArchiverInOut
{

    private SerializableCompositeFactory compositeFactory;
    private ReadByteBuffer readBuf;

    public OPSArchiverIn(ReadByteBuffer buf)
    {
        compositeFactory = OPSObjectFactory.getInstance();
        readBuf = buf;
    }

    public boolean isOut() { return false; }

    public boolean remove(Object o)
    {
        return compositeFactory.remove(o);
    }

    public boolean add(SerializableFactory e)
    {
        return compositeFactory.add(e);
    }
    public int inout(String name, int v) throws IOException
    {
        return readBuf.readint();
    }

    public long inout(String name, long v) throws IOException
    {
        return readBuf.readlong();
    }

    public byte inout(String name, byte v) throws IOException
    {
        return readBuf.readbyte();
    }

    public short inout(String name, short v) throws IOException
    {
        return readBuf.readshort();
    }

    public float inout(String name, float v) throws IOException
    {
        return readBuf.readfloat();
    }

    public boolean inout(String name, boolean v) throws IOException
    {
        return readBuf.readboolean();
    }

    public String inout(String name, String v) throws IOException
    {
        return readBuf.readstring();
    }

    public double inout(String name, double v) throws IOException
    {
        return readBuf.readdouble();
    }

    public <T extends Serializable> Serializable inout(String name, Serializable v, Class<T> cls) throws IOException
    {
        String type = readBuf.readstring();     // Skip string for non virtual objects, type is known by T
        Serializable newSer = null;
        try {
          newSer = (Serializable) cls.newInstance();
          newSer.serialize(this);
        } catch (InstantiationException | IllegalAccessException e)
        {
        }
        return newSer;
    }

    public Serializable inout(String name, Serializable v) throws IOException
    {
        String type = readBuf.readstring();
        Serializable newSer = compositeFactory.create(type);
        if(newSer != null)
        {
            newSer.serialize(this);
        }
        return newSer;
    }

    public List<Integer> inoutIntegerList(String name, List<Integer> v) throws IOException
    {
        return (List<Integer>) readBuf.readintArr();
    }

    public List<Long> inoutLongList(String name, List<Long> v) throws IOException
    {
        return (List<Long>) readBuf.readlongArr();
    }

    public List<Byte> inoutByteList(String name, List<Byte> v) throws IOException
    {
        return (List<Byte>) readBuf.readbyteArr();
    }

    public List<Short> inoutShortList(String name, List<Short> v) throws IOException
    {
        return (List<Short>) readBuf.readshortArr();
    }

    public List<Float> inoutFloatList(String name, List<Float> v) throws IOException
    {
        return (List<Float>) readBuf.readfloatArr();
    }

    public List<Boolean> inoutBooleanList(String name, List<Boolean> v) throws IOException
    {
         return (List<Boolean>) readBuf.readbooleanArr();
    }

    public List<String> inoutStringList(String name, List<String> v) throws IOException
    {
         return (List<String>) readBuf.readstringArr();
    }

    public List<Double> inoutDoubleList(String name, List<Double> v) throws IOException
    {
         return (List<Double>) readBuf.readdoubleArr();
    }

    public List inoutSerializableList(String name, List v) throws IOException
    {
        Vector list = new Vector();
        int size = readBuf.readint();
        for (int i = 0; i < size; i++)
        {
            list.add(inout("", (Serializable)null));
        }
        return list;
    }

    public <T extends Serializable> List inoutSerializableList(String name, List v, Class<T> cls) throws IOException
    {
        Vector list = new Vector();
        int size = readBuf.readint();
        for (int i = 0; i < size; i++)
        {
            list.add(inout("", (Serializable)null, cls));
        }
        return list;
    }

}
